packages/compile-solidity/compileWithPragmaAnalysis.js

```diff
@@ -0,0 +1,139 @@
+    if (semver.validRange(expression) === null) {
+      const message = `Invalid semver expression (${expression}) found in` +
+        `one of your contract's imports.`;
+      throw new Error(message);
+    }
+  }
+};
+
+// takes an array of versions and an array of semver expressions
+// returns a version of the compiler or undefined if none can be found
+const findNewestSatisfyingVersion = ({ solcReleases, semverExpressions }) => {
+  // releases are ordered from newest to oldest
+  return solcReleases.find(version => {
+    return semverExpressions.every(expression =>
+      semver.satisfies(version, expression)
+    );
+  });
+};
+
+const throwCompilerVersionNotFound = ({ path, semverExpressions }) => {
+  const message =
+    `Could not find a single version of the Solidity compiler that ` +
+    `satisfies the following semver expressions obtained from your source ` +
+    `files' pragma statements: ${semverExpressions.join(" - ")}. ` +
+    `${OS.EOL}Please check the pragma statements for ${path} and its imports.`;
+  throw new Error(message);
+};
+
+const compileWithPragmaAnalysis = async ({ paths, options }) => {
+  const supplierOptions = {
+    events: options.events,
+    solcConfig: options.compilers.solc
+  };
+  const compilerSupplier = new CompilerSupplier(supplierOptions);
+  const { releases } = await compilerSupplier.getReleases();
+
+  // collect sources by the version of the Solidity compiler that they require
+  const versionsAndSources = {};
+  for (const path of paths) {
+    const source = (await options.resolver.resolve(path)).body;
+
+    const parserVersion = findNewestSatisfyingVersion({
+      solcReleases: releases,
+      semverExpressions: [getSemverExpression(source)]
+    });
+    if (!parserVersion) {
+      throwCompilerVersionNotFound({
+        path,
+        semverExpressions: [getSemverExpression(source)]
+      });
+    }
+
+    // allSources is of the format { [filename]: string }
+    const { allSources } = await Profiler.requiredSourcesForSingleFile(
+      options.with({
+        path,
+        base_path: options.contracts_directory,
+        resolver: options.resolver,
+        compilers: {
+          solc: {
+            version: parserVersion
+          }
+        }
+      })
+    );
+
+    // get an array of all the semver expressions in the sources
+    const semverExpressions = await getSemverExpressions(
+      Object.values(allSources)
+    );
+
+    // this really just validates the expressions from the contracts' imports
+    // as it has already determined the parser version for each contract
+    validateSemverExpressions(semverExpressions);
+
+    const newestSatisfyingVersion = findNewestSatisfyingVersion({
+      solcReleases: releases,
+      semverExpressions
+    });
+    if (!newestSatisfyingVersion) {
+      throwCompilerVersionNotFound({
+        path,
+        semverExpressions
+      });
+    }
+
+    if (versionsAndSources[newestSatisfyingVersion]) {
```

Comment:
You actually don't need a conditional here!  Inside an object literal (note: *not* in any other context!), it's actually legal to do `...undefined`!  It acts the same as `...{}`.

Reply:
Oh cool